lib/rspec/matchers/built_in/include.rb

```diff
@@ -51,27 +67,50 @@ def diffable?
 
       private
 
+        def readable_list_of(items)
+          described_items = surface_descriptions_in(items)
+          if described_items.all? { |item| item.is_a?(Hash) }
+            " " + described_items.inject(:merge).to_s
+          else
+            EnglishPhrasing.list(described_items)
+          end
+        end
+
         def invalid_object_message
-          return '' if actual.respond_to?(:include?)
           ", but it does not respond to `include?`"
         end
 
-        def perform_match(predicate, hash_subset_predicate)
-          return false unless actual.respond_to?(:include?)
-
-          expected.__send__(predicate) do |expected_item|
+        def excluded_from_actual(options={ :invert => false })
+          expected.inject([]) do |memo, expected_item|
             if comparing_hash_to_a_subset?(expected_item)
-              expected_item.__send__(hash_subset_predicate) do |(key, value)|
-                actual_hash_includes?(key, value)
+              expected_item.each do |(key, value)|
+                if options[:invert] == true
+                  memo << { key => value } if actual_hash_includes?(key, value)
+                else
+                  memo << { key => value } unless actual_hash_includes?(key, value)
+                end
               end
             elsif comparing_hash_keys?(expected_item)
-              actual_hash_has_key?(expected_item)
+              if options[:invert] == true
+                memo << expected_item if actual_hash_has_key?(expected_item)
+              else
+                memo << expected_item unless actual_hash_has_key?(expected_item)
+              end
             else
-              actual_collection_includes?(expected_item)
+              if options[:invert] == true
+                memo << expected_item if actual_collection_includes?(expected_item)
+              else
+                memo << expected_item unless actual_collection_includes?(expected_item)
+              end
             end
+            memo
```

Comment:
This method checks the same conditional 3 times and, IMO, is crying out for a more polymorphic approach.  Here's how we can do that, I think:

``` ruby
class Include
  def matches?(actual)
    perform_match(actual) { |v| v }
  end

  def does_not_match?(actual)
    perform_match(actual) { |v| !v }
  end

private

  def perform_match(actual, &block)
    @actual = actual
    @divergent_items = excluded_from_actual(&block)
    actual.respond_to?(:include?) && @divergent_items.include?
  end

  def excluded_from_actual
    return [] unless @actual.respond_to?(:include?)

    expected.inject([]) do |memo, expected_item|
      if comparing_hash_to_a_subset?(expected_item)
        expected_item.each do |(key, value)|
          memo << { key => value } unless (yield actual_hash_includes?(key, value))
        end
      elsif comparing_hash_keys?(expected_item)
        memo << expected_item unless (yield actual_hash_has_key?(expected_item))
      else
        memo << expected_item unless (yield actual_collection_includes?(expected_item))
      end
      memo
    end
  end
end
```

Thoughts? (This also further improves the `matches?` and `does_not_match?` suggestions from above).